core/src/trezor/wire/thp/received_message_handler.py

```diff
@@ -0,0 +1,389 @@
+        raise ThpError("Message received is not a handshake init request!")
+    if not payload_length == PUBKEY_LENGTH + CHECKSUM_LENGTH:
+        raise ThpError("Message received is not a valid handshake init request!")
+
+    ctx.handshake = Handshake()
+
+    host_ephemeral_pubkey = bytearray(
+        ctx.buffer[INIT_HEADER_LENGTH : message_length - CHECKSUM_LENGTH]
+    )
+    trezor_ephemeral_pubkey, encrypted_trezor_static_pubkey, tag = (
+        ctx.handshake.handle_th1_crypto(
+            thp_messages.get_encoded_device_properties(), host_ephemeral_pubkey
+        )
+    )
+
+    if __debug__:
+        log.debug(
+            __name__,
+            "trezor ephemeral pubkey: %s",
+            hexlify(trezor_ephemeral_pubkey).decode(),
+        )
+        log.debug(
+            __name__,
+            "trezor masked static pubkey: %s",
+            hexlify(encrypted_trezor_static_pubkey).decode(),
+        )
+        log.debug(__name__, "tag: %s", hexlify(tag))
+
+    payload = trezor_ephemeral_pubkey + encrypted_trezor_static_pubkey + tag
+
+    # send handshake init response message
+    await ctx.write_handshake_message(HANDSHAKE_INIT_RES, payload)
+    ctx.set_channel_state(ChannelState.TH2)
+    return
+
+
+async def _handle_state_TH2(ctx: Channel, message_length: int, ctrl_byte: int) -> None:
+    from apps.thp.credential_manager import validate_credential
+
+    if __debug__:
+        log.debug(__name__, "handle_state_TH2")
+    if not control_byte.is_handshake_comp_req(ctrl_byte):
+        raise ThpError("Message received is not a handshake completion request!")
+    if ctx.handshake is None:
+        raise Exception("Handshake object is not prepared. Retry handshake.")
+
+    host_encrypted_static_pubkey = memoryview(ctx.buffer)[
+        INIT_HEADER_LENGTH : INIT_HEADER_LENGTH + KEY_LENGTH + TAG_LENGTH
+    ]
+    handshake_completion_request_noise_payload = memoryview(ctx.buffer)[
+        INIT_HEADER_LENGTH + KEY_LENGTH + TAG_LENGTH : message_length - CHECKSUM_LENGTH
+    ]
+
+    ctx.handshake.handle_th2_crypto(
+        host_encrypted_static_pubkey, handshake_completion_request_noise_payload
+    )
+
+    ctx.channel_cache.set(CHANNEL_KEY_RECEIVE, ctx.handshake.key_receive)
+    ctx.channel_cache.set(CHANNEL_KEY_SEND, ctx.handshake.key_send)
+    ctx.channel_cache.set(CHANNEL_HANDSHAKE_HASH, ctx.handshake.h)
+    ctx.channel_cache.set_int(CHANNEL_NONCE_RECEIVE, 0)
+    ctx.channel_cache.set_int(CHANNEL_NONCE_SEND, 1)
+
+    noise_payload = thp_messages.decode_message(
+        ctx.buffer[
+            INIT_HEADER_LENGTH
+            + KEY_LENGTH
+            + TAG_LENGTH : message_length
+            - CHECKSUM_LENGTH
+            - TAG_LENGTH
+        ],
+        0,
+        "ThpHandshakeCompletionReqNoisePayload",
+    )
+    if TYPE_CHECKING:
+        assert ThpHandshakeCompletionReqNoisePayload.is_type_of(noise_payload)
+    for i in noise_payload.pairing_methods:
```

Comment:
You should check that `noise_payload.pairing_methods` is a subset of `_ENABLED_PAIRING_METHODS`.

Reply:
verification implemented here: https://github.com/trezor/trezor-firmware/pull/4050/commits/1f7ffd2edabe713963c20c527659d7d43acd77a9

Reply:
I am wondering if it might be better to raise an exception if `noise_payload.pairing_methods` is not a subset of the enabled pairing methods. Lets discuss it on Monday.

Reply:
exception + error message implemented here:
https://github.com/trezor/trezor-firmware/pull/4050/commits/e4029341cfae750c82ced7ffa7cf3d7410971a99